clients/include/testing_csx2dense.hpp

```diff
@@ -0,0 +1,396 @@
+
+    auto m_csx_row_col_ptr
+        = hipsparse_unique_ptr{device_malloc(sizeof(int) * (DIMDIR + 1)), device_free};
+    auto m_csx_val         = hipsparse_unique_ptr{device_malloc(sizeof(T) * nnz), device_free};
+    auto m_csx_col_row_ind = hipsparse_unique_ptr{device_malloc(sizeof(int) * nnz), device_free};
+
+    int* d_csx_row_col_ptr = (int*)m_csx_row_col_ptr.get();
+    int* d_csx_col_row_ind = (int*)m_csx_col_row_ind.get();
+    T*   d_csx_val         = (T*)m_csx_val.get();
+
+    if(!d_csx_row_col_ptr || !d_csx_val || !d_csx_col_row_ind)
+    {
+        CHECK_HIP_ERROR(hipErrorOutOfMemory);
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
+    }
+
+    std::vector<int> cpu_csx_row_col_ptr(DIMDIR + 1);
+    std::vector<T>   cpu_csx_val(nnz);
+    std::vector<int> cpu_csx_col_row_ind(nnz);
+    if(!cpu_csx_row_col_ptr.data() || !cpu_csx_val.data() || !cpu_csx_col_row_ind.data())
+    {
+        CHECK_HIP_ERROR(hipErrorOutOfMemory);
+        return HIPSPARSE_STATUS_ALLOC_FAILED;
+    }
+
+    //
+    // Convert the dense matrix to a compressed sparse matrix.
+    //
+    if(DIRA == HIPSPARSE_DIRECTION_ROW)
+    {
+        CHECK_HIPSPARSE_ERROR(dense2csx(handle,
+                                        M,
+                                        N,
+                                        descr,
+                                        (const T*)d_dense_val,
+                                        LD,
+                                        d_nnzPerRowColumn,
+                                        d_csx_val,
+                                        d_csx_row_col_ptr,
+                                        d_csx_col_row_ind));
+    }
+    else
+    {
+        CHECK_HIPSPARSE_ERROR(dense2csx(handle,
+                                        M,
+                                        N,
+                                        descr,
+                                        (const T*)d_dense_val,
+                                        LD,
+                                        d_nnzPerRowColumn,
+                                        d_csx_val,
+                                        d_csx_col_row_ind,
+                                        d_csx_row_col_ptr));
+    }
+
+    //
+    // Copy on host.
+    //
+    CHECK_HIP_ERROR(hipMemcpy(
+        cpu_csx_val.data(), d_csx_val, sizeof(T) * std::max(nnz, 1), hipMemcpyDeviceToHost));
+
+    CHECK_HIP_ERROR(hipMemcpy(cpu_csx_row_col_ptr.data(),
+                              d_csx_row_col_ptr,
+                              sizeof(int) * (DIMDIR + 1),
+                              hipMemcpyDeviceToHost));
+
+    CHECK_HIP_ERROR(hipMemcpy(cpu_csx_col_row_ind.data(),
+                              d_csx_col_row_ind,
+                              sizeof(int) * std::max(nnz, 1),
+                              hipMemcpyDeviceToHost));
+
+    if(argus.unit_check)
+    {
+        for(int i = 0; i < LD; ++i)
```

Comment:
Again collapse into single loop if you want.

Reply:
In fact I think h_dense_val.resize(N*LD, static_cast<T>(-2)); would work

Reply:
sure